File: lib/utils/delivery_schedule.dart
import 'package:delycafe/utils/app_timezone.dart';

/// Правила приёма заказов (время кафе: Asia/Yekaterinburg).
class DeliverySchedule {
  DeliverySchedule._();

  static const int openHour = 10;
  static const int openMinute = 0;
  static const int slotIntervalMinutes = 5;

  static const int sundayThroughThursdayCloseHour = 20;
  static const int sundayThroughThursdayCloseMinute = 30;
  static const int fridaySaturdayCloseHour = 21;
  static const int fridaySaturdayCloseMinute = 30;

  /// Текст для UI: вс–чт до 20:30, пт–сб до 21:30.
  static const String acceptanceHoursShort =
      'вс–чт до 20:30, пт–сб до 21:30';

  static const String acceptanceHoursLong =
      'Приём заказов: с 10:00, вс–чт до 20:30, пт–сб до 21:30';

  static DateTime get now => cafeNow();

  static Duration leadTime(String deliveryTypeApi) {
    if (deliveryTypeApi == 'tatysh') {
      return const Duration(hours: 2);
    }

    return const Duration(hours: 1, minutes: 30);
  }

  /// Приём заказов открыт: с 10:00 до 20:30 (вс–чт) / 21:30 (пт–сб).
  static bool isAcceptingOrders(DateTime now) {
    return _isWithinKitchenHours(now);
  }

  static bool isAnyOrderingOpen(DateTime now) {
    return isAcceptingOrders(now);
  }

  /// Есть ли сегодня слот доставки с учётом lead time.
  static bool isOrderingOpen(DateTime now, String deliveryTypeApi) {
    if (!isAcceptingOrders(now)) {
      return false;
    }

    return _canDeliverToday(now, deliveryTypeApi);
  }

  /// До 10:00 или с 20:30 (вс–чт) / 21:30 (пт–сб) — кухня закрыта.
  static bool isKitchenClosed(DateTime now) {
    return !isAcceptingOrders(now);
  }

  static bool _isWithinKitchenHours(DateTime now) {
    if (now.isBefore(_openOnDate(now))) {
      return false;
    }

    return now.isBefore(_closingTimeOnDate(now));
  }

  static bool _canDeliverToday(DateTime now, String deliveryTypeApi) {
    final minDelivery = minDeliveryTime(now, deliveryTypeApi);
    final maxDelivery = maxDeliveryTime(now);

    return !minDelivery.isAfter(maxDelivery);
  }

  static DateTime nextOrderingOpensAt(DateTime now) {
    final todayOpen = _openOnDate(now);

    if (now.isBefore(todayOpen)) {
      return todayOpen;
    }

    final nextDay = cafeDateTime(now.year, now.month, now.day).add(
      const Duration(days: 1),
    );

    return _openOnDate(nextDay);
  }

  static DateTime minDeliveryTime(DateTime now, String deliveryTypeApi) {
    final todayOpen = _openOnDate(now);
    final earliest = now.add(leadTime(deliveryTypeApi));

    if (earliest.isBefore(todayOpen)) {
      return todayOpen;
    }

    return earliest;
  }

  static DateTime maxDeliveryTime(DateTime now) {
    return _closingTimeOnDate(now);
  }

  static List<DateTime> availableSlots(DateTime now, String deliveryTypeApi) {
    if (!isAcceptingOrders(now)) {
      return const [];
    }

    final minSlot = _roundUpToInterval(
      minDeliveryTime(now, deliveryTypeApi),
      slotIntervalMinutes,
    );
    final maxSlot = maxDeliveryTime(now);

    if (minSlot.isAfter(maxSlot)) {
      return [maxSlot];
    }

    final slots = <DateTime>[];
    var cursor = minSlot;

    while (!cursor.isAfter(maxSlot)) {
      slots.add(cursor);
      cursor = cursor.add(
        const Duration(minutes: slotIntervalMinutes),
      );
    }

    return slots;
  }

  static String formatTime(DateTime time) {
    final hour = time.hour.toString().padLeft(2, '0');
    final minute = time.minute.toString().padLeft(2, '0');

    return '$hour:$minute';
  }

  static String formatNextOpening(DateTime now) {
    final opensAt = nextOrderingOpensAt(now);
    final today = cafeDateTime(now.year, now.month, now.day);
    final opensDay = cafeDateTime(opensAt.year, opensAt.month, opensAt.day);

    if (opensDay == today) {
      return 'сегодня в ${formatTime(opensAt)}';
    }

    final tomorrow = today.add(const Duration(days: 1));

    if (opensDay == tomorrow) {
      return 'завтра в ${formatTime(opensAt)}';
    }

    return '${opensAt.day.toString().padLeft(2, '0')}.'
        '${opensAt.month.toString().padLeft(2, '0')} '
        'в ${formatTime(opensAt)}';
  }

  static String closedMessage(DateTime now) {
    return 'Приём заказов закрыт. Оформление снова будет доступно '
        '${formatNextOpening(now)}.';
  }

  static DateTime firstSlotAfterOpen(
    DateTime openAt,
    String deliveryTypeApi,
  ) {
    final earliest = openAt.add(leadTime(deliveryTypeApi));

    return _roundUpToInterval(earliest, slotIntervalMinutes);
  }

  static DateTime displayAsapSlot(DateTime now, String deliveryTypeApi) {
    if (isAcceptingOrders(now)) {
      final earliest = minDeliveryTime(now, deliveryTypeApi);
      final latest = maxDeliveryTime(now);
      final target = earliest.isAfter(latest) ? latest : earliest;

      return _roundUpToInterval(
        target,
        slotIntervalMinutes,
      );
    }

    return firstSlotAfterOpen(
      nextOrderingOpensAt(now),
      deliveryTypeApi,
    );
  }

  static String asapChoiceLabel(DateTime now, String deliveryTypeApi) {
    final slot = displayAsapSlot(now, deliveryTypeApi);
    final time = formatTime(slot);

    if (isAcceptingOrders(now)) {
      return 'к $time';
    }

    return 'завтра к $time';
  }

  static String previewTimeLabel(DateTime now, String deliveryTypeApi) {
    return formatTime(displayAsapSlot(now, deliveryTypeApi));
  }

  static bool isClosedUntilTomorrow(DateTime now) {
    if (now.isBefore(_openOnDate(now))) {
      return false;
    }

    return true;
  }

  static String closedSubmitButtonLabel(DateTime now) {
    return 'Кухня закрыта';
  }

  static String asapEstimateMessage(DateTime now, String deliveryTypeApi) {
    return 'Ориентировочно ${asapChoiceLabel(now, deliveryTypeApi)}';
  }

  static int _lastSlotHour(int weekday) {
    if (_isFridayOrSaturday(weekday)) {
      return fridaySaturdayCloseHour;
    }

    return sundayThroughThursdayCloseHour;
  }

  static int _lastSlotMinute(int weekday) {
    if (_isFridayOrSaturday(weekday)) {
      return fridaySaturdayCloseMinute;
    }

    return sundayThroughThursdayCloseMinute;
  }

  static bool _isFridayOrSaturday(int weekday) {
    return weekday == DateTime.friday || weekday == DateTime.saturday;
  }

  static DateTime _openOnDate(DateTime date) {
    return cafeDateTime(
      date.year,
      date.month,
      date.day,
      openHour,
      openMinute,
    );
  }

  static DateTime _closingTimeOnDate(DateTime date) {
    return cafeDateTime(
      date.year,
      date.month,
      date.day,
      _lastSlotHour(date.weekday),
      _lastSlotMinute(date.weekday),
    );
  }

  static DateTime _roundUpToInterval(DateTime value, int intervalMinutes) {
    final totalMinutes = value.hour * 60 + value.minute;
    final remainder = totalMinutes % intervalMinutes;

    if (remainder == 0 &&
        value.second == 0 &&
        value.millisecond == 0 &&
        value.microsecond == 0) {
      return cafeDateTime(
        value.year,
        value.month,
        value.day,
        value.hour,
        value.minute,
      );
    }

    final roundedMinutes = totalMinutes + (intervalMinutes - remainder);

    return cafeDateTime(value.year, value.month, value.day).add(
      Duration(minutes: roundedMinutes),
    );
  }
}

File: backend/orders/delivery_schedule.py
from datetime import date, datetime, time, timedelta

from django.utils import timezone

from orders.models import Order

OPEN_TIME = time(10, 0)
# Вс–чт: приём заказов до 20:30, пт–сб: до 21:30.
SLOT_SUNDAY_THROUGH_THURSDAY = time(20, 30)
SLOT_FRIDAY_SATURDAY = time(21, 30)
LEAD_DEFAULT = timedelta(hours=1, minutes=30)
LEAD_TATYSH = timedelta(hours=2)
SLOT_INTERVAL = timedelta(minutes=5)


def lead_for_delivery_type(delivery_type: str) -> timedelta:
    if delivery_type == Order.DeliveryType.TATYSH:
        return LEAD_TATYSH

    return LEAD_DEFAULT


def last_slot_for_date(day: date) -> time:
    # Python weekday: пн=0 … вс=6. Пт=4, сб=5 — до 21:30, вс–чт — до 20:30.
    if day.weekday() in (4, 5):
        return SLOT_FRIDAY_SATURDAY

    return SLOT_SUNDAY_THROUGH_THURSDAY


def combine_local(day: date, value: time) -> datetime:
    return timezone.make_aware(datetime.combine(day, value))


def _round_up_to_interval(value: datetime, interval_minutes: int = 5) -> datetime:
    local_value = timezone.localtime(value)
    total_minutes = local_value.hour * 60 + local_value.minute
    remainder = total_minutes % interval_minutes

    if remainder == 0 and local_value.second == 0 and local_value.microsecond == 0:
        return local_value.replace(second=0, microsecond=0)

    rounded_minutes = total_minutes + (interval_minutes - remainder)

    base = local_value.replace(hour=0, minute=0, second=0, microsecond=0)
    return base + timedelta(minutes=rounded_minutes)


def min_delivery_datetime(now: datetime, delivery_type: str) -> datetime:
    local_now = timezone.localtime(now)
    today = local_now.date()
    open_dt = combine_local(today, OPEN_TIME)
    earliest = local_now + lead_for_delivery_type(delivery_type)

    if earliest < open_dt:
        return open_dt

    return earliest


def max_delivery_datetime(now: datetime) -> datetime:
    local_now = timezone.localtime(now)
    today = local_now.date()

    return combine_local(today, last_slot_for_date(today))


def is_kitchen_closed(now: datetime) -> bool:
    local_now = timezone.localtime(now)
    today = local_now.date()
    open_dt = combine_local(today, OPEN_TIME)
    close_dt = combine_local(today, last_slot_for_date(today))

    if local_now < open_dt:
        return True

    return local_now >= close_dt


def is_ordering_open(now: datetime, delivery_type: str) -> bool:
    """Приём заказов открыт по расписанию кухни (без учёта lead time)."""
    return not is_kitchen_closed(now)


def has_delivery_slots_today(now: datetime, delivery_type: str) -> bool:
    if is_kitchen_closed(now):
        return False

    local_now = timezone.localtime(now)
    min_dt = min_delivery_datetime(local_now, delivery_type)
    max_dt = max_delivery_datetime(local_now)

    return min_dt <= max_dt


def next_ordering_opens_at(now: datetime) -> datetime:
    local_now = timezone.localtime(now)
    today = local_now.date()
    open_today = combine_local(today, OPEN_TIME)

    if local_now < open_today:
        return open_today

    tomorrow = today + timedelta(days=1)
    return combine_local(tomorrow, OPEN_TIME)


def available_slot_times(now: datetime, delivery_type: str) -> list[str]:
    if is_kitchen_closed(now):
        return []

    local_now = timezone.localtime(now)
    min_slot = _round_up_to_interval(min_delivery_datetime(local_now, delivery_type))
    max_slot = max_delivery_datetime(local_now)

    if min_slot > max_slot:
        return [max_slot.strftime('%H:%M')]

    slots: list[str] = []
    cursor = min_slot

    while cursor <= max_slot:
        slots.append(cursor.strftime('%H:%M'))
        cursor += SLOT_INTERVAL

    return slots


def validate_order_delivery_window(
    *,
    now: datetime,
    delivery_type: str,
    delivery_time_type: str,
    delivery_time: str,
) -> None:
    from rest_framework import serializers

    if is_kitchen_closed(now):
        opens_at = next_ordering_opens_at(now)
        raise serializers.ValidationError(
            'Приём заказов сейчас закрыт. '
            f'Оформление снова будет доступно с {opens_at.strftime("%H:%M")}.'
        )

    if delivery_time_type != Order.DeliveryTimeType.BY_TIME:
        return

    allowed = set(available_slot_times(now, delivery_type))
    normalized_time = (delivery_time or '').strip()

    if normalized_time not in allowed:
        raise serializers.ValidationError(
            'Выбранное время доставки недоступно. Обновите время и попробуйте снова.'
        )
